numalogic/blocks/pipeline.py

```diff
@@ -0,0 +1,173 @@
+# Copyright 2022 The Numaproj Authors.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections.abc import Sequence
+from collections.abc import Iterator
+
+import numpy.typing as npt
+
+from numalogic.blocks._transform import Block
+from numalogic.registry import ArtifactManager
+from numalogic.tools.types import artifact_t
+
+
+class BlockPipeline(Sequence[Block]):
```

Comment:
It seems one pipeline instance is used to hold and use one model. To host multiple models, do you instantiate/load multiple pipelines, one per key? Do you need a router component, that picks the right pipeline based on the key? Should an optimization be added, where the stateless blocks are reused, and for the stateful blocks just the artifacts are cached, and again not instantiated with every model?

Reply:
Yes, for now the approach would be to instantiate multiple pipelines for different composite key models. Soon, let's optimize this so that stateless blocks are reused. We probably need to design it in a way that it is threadsafe, while not taking too much memory.